### ISSUE TITLE
Make all pipeline steps opt-in

### ISSUE THREAD

**Issue by prototaxites** [nf-core/mag#455]
### Description of feature

At the moment, a default pipeline run entails a highly fully-featured run with many optional steps enabled that may or may not be desirable - for example, assembly will happen with Megahit and SPAdes, and SPAdesHybrid (if long reads are supplied), and all three binning algorithms will run. All post-binning steps are run as well, which means for a highly forking analysis that covers all of the potential tool combination space. All of these have to be explicitly disabled in order to run a more minimal pipeline, and it can be easy to forget to turn one off.

To reduce the footprint of a default run, all steps could be opt-in, replacing `--skip_{tool}` with `--enable_{tool}` calls to turn them on. This would be cleaner than some of the `--skip_{tool}` options, which are not uniformly used throughout the pipeline (especially where database inputs are also required for a tool to run - for example, CAT runs if a database is supplied, but does not have a dedicated skip parameter), and might also have the advantage that users have to think more carefully about the analysis they want to run before hitting go.

**Comment by d4straub** [comment]
I am feel ambivalent here. The original idea of the pipeline was to have by default a comprehensive analysis. However, the original version had also much less tools available. I do like to have with default parameters a comprehensive analysis, because it will make it much easier to run the pipeline, with minimal additional parameters (i.e. in my opinion minimizing what can be forgotten to specify). Additionally, it comes with an easy start for users, because it minimizes the required investment in reading the documentation. What you are proposing seems to be the opposite, i.e. first read the documentation carefully, understand all that, then start the pipeline with a plethora of parameters.
I do like the idea of having a single param for supplying DBs to trigger a tool to run, but I agree that this is currently not perfectly solved in every case.
Having said that, I am obviously only defending what I have broken (and what I produced primarily for my own needs and preferences)...

**Comment by prototaxites** [comment]
Fair points - I definitely understand the desire to have a default configuration that runs a comprehensive analysis - we can easily run into the issue that users don't know which parts of the pipeline to turn on or off. Personally, I shy away from defaults unless I understand what they are doing, as they can be come across as 'one size fits all' when there are caveats that need to be considered in each use-case, which might lead to wrong inferences downstream.

One option that occurs to me, were we to make these changes, would be to create a set of profiles similar to the test profiles, that enable or disable different sets of options, while still requiring inputs such as samplesheets. Those could be things like "minimal", "fast", "everything", "prokarya" and specified with `-profile`. That would enable a flexible pipeline for those who need it while helping new users to choose from a set of configurations that suit their needs.

**Comment by jfy133** [comment]
I think conceptually additional profiles in this regard makes sense. I've already done something similar with `-profile test_nothing` :grimacing: .

But for users: I vaguely remember there was some concerns about the use of profiles (and overwriting issues, if users supply their own). It might be worth checking on nf-core for that.

**Comment by prototaxites** [comment]
Perhaps I've missed a complexity, but I would have thought that users supplying custom configs would not be a problem - custom configs take precedence over the parameters selected in `nextflow.config` (and hence profiles specified there?), so calling a preset configuration alongside a custom config would just create a novel configuration that maybe adds or removes particular steps, the same as calling parameters from the command line. The only obvious issue I can see occurring would be parameter conflicts (with regards to those specified in `WorkflowMag.groovy`), but these might be easier to avoid with an explicit `--enable-{tool}` for every process.

But I appreciate that this may not be an "nf-core" way of doing things - definitely worth checking!

**Comment by prototaxites** [comment]
I followed up with this on Slack: https://nfcore.slack.com/archives/CE6SDBX2A/p1687424806436439

The conclusion seems to be that there shouldn't be a problem supplying pre-set pipeline configs as profiles, as long as we're careful to specify incompatible parameter options within the pipeline and throw an error when they're met. Supplying parameters as a YML/JSON file, in a custom config, or on the command line should all supercede a parameter set in a specified profile, following the specification in the [Nextflow docs](https://www.nextflow.io/docs/latest/config.html). The only other thing to avoid would be config cascading, where there are multiple nested levels of `includeConfig {}`, and to only include configs within `nextflow.config`.

We could also be a bit cagey and outright state that when using pre-set configs, use of a custom config file to change parameter options is not recommended, and could lead to hard-to-reproduce behaviour in rare cases. That said, there will be some parameters that many users are going to have to set, mostly those relating to databases: `cat_db`, `gtdb`, `busco_reference`/`busco_auto_lineage`, etc., so we should make sure that work with all profiles when changed if we take this route.

**Comment by d4straub** [comment]
> cat_db, gtdb, busco_reference/busco_auto_lineage, etc., so we should make sure that work with all profiles when changed if we take this route.

are anyway not recommended to set within a config but have to be set via `-params-file`, which is to my knowledge (that is incomplete) overwriting any `.config` setting in a clean way, similar to parameters specified via the command line.

**Comment by alneberg** [comment]
Jumping into this discussion. Since I prefer to use the form-style of starting pipelines (previously with nf-core launch, and now instead with seqera platform) I think the current way the form works would be hard to combine with these type of profiles. While I like the idea I think it's very important to be able to start from one of these pre-defined profiles and still be able to modify some parameters, in my case I would like to do that in a launch form in seqera platform. It could of course be possible to adapt seqera platform to this bevaviour, but I don't know how hard that would be.

Supplying several predefined parameter files would integrate much easier I think since those one can upload and they will be used to pre-fill the form for a user.

**Comment by jfy133** [comment]
@alneberg so you prefer keeping a default set of functionality turned on, and make them skippable?

**Comment by alneberg** [comment]
Yes exactly. It's not a very strong negative opinion though against the profiles.
